rust/lance/src/dataset/write/update.rs

```diff
@@ -0,0 +1,566 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{BTreeMap, HashMap};
+use std::sync::{Arc, RwLock};
+
+use super::super::utils::make_rowid_capture_stream;
+use super::write_fragments_internal;
+use arrow_array::{Array, ArrayRef, FixedSizeListArray, RecordBatch};
+use arrow_schema::{ArrowError, DataType, Schema as ArrowSchema};
+use arrow_select::concat::concat;
+use datafusion::common::DFSchema;
+use datafusion::error::{DataFusionError, Result as DFResult};
+use datafusion::logical_expr::ExprSchemable;
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+use datafusion::physical_plan::{ColumnarValue, PhysicalExpr};
+use datafusion::prelude::Expr;
+use datafusion::scalar::ScalarValue;
+use futures::StreamExt;
+use lance_arrow::RecordBatchExt;
+use lance_core::error::{box_error, InvalidInputSnafu};
+use lance_core::format::Fragment;
+use lance_index::util::datafusion::safe_coerce_scalar;
+use roaring::RoaringTreemap;
+use snafu::{location, Location, ResultExt};
+
+use crate::dataset::transaction::{Operation, Transaction};
+use crate::io::commit::commit_transaction;
+use crate::{io::exec::Planner, Dataset};
+use crate::{Error, Result};
+
+/// Update a dataset
+///
+/// ```ignore
+/// let dataset = UpdateBuilder::new()
+///     .update_where("region_id = 10")
+///     .set("region_name", "New York")
+///     .build?
+///     .execute()
+///     .await?;
+/// ```
+///
+#[derive(Debug, Clone)]
+pub struct UpdateBuilder {
+    /// The dataset snapshot to update.
+    dataset: Arc<Dataset>,
+    /// The condition to apply to find matching rows to update. If None, all rows are updated.
+    condition: Option<Expr>,
+    /// The updates to apply to matching rows.
+    updates: HashMap<String, Expr>,
+}
+
+impl UpdateBuilder {
+    pub fn new(dataset: Arc<Dataset>) -> Self {
+        Self {
+            dataset,
+            condition: None,
+            updates: HashMap::new(),
+        }
+    }
+
+    pub fn update_where(mut self, filter: &str) -> Result<Self> {
+        let planner = Planner::new(Arc::new(self.dataset.schema().into()));
+        let expr = planner
+            .parse_filter(filter)
+            .map_err(box_error)
```

Comment:
What is `map_err(box_error)`?

Reply:
`box_error` is just `Box::new()` with a type annotation. I would if I could just use `Box::new` here...

Reply:
Turns out it needs to stay as `box_error`.